src/controllers/PartialUserFormController.php

```diff
@@ -45,54 +49,57 @@ public function partial(HTTPRequest $request)
 
         $key = $request->param('Key');
         $token = $request->param('Token');
+        if (!$key || !$token) {
+            return $this->httpError(404);
+        }
 
+        /** @var PartialFormSubmission $partial */
         $partial = PartialFormSubmission::get()->find('Token', $token);
-        if (!$token || !$partial || !$partial->UserDefinedFormID) {
+        if (!$partial ||
+            !$partial->UserDefinedFormID ||
+            !hash_equals($partial->generateKey($token), $key)
+        ) {
             return $this->httpError(404);
         }
 
-        // TODO: Recognize visitor with the password
-        if ($partial->generateKey($token) === $key) {
-            // Set the session if the last session has expired
-            if (!$request->getSession()->get(PartialSubmissionController::SESSION_KEY)) {
-                $request->getSession()->set(PartialSubmissionController::SESSION_KEY, $partial->ID);
-            }
+        // Set the session if the last session has expired
+        if (!$request->getSession()->get(PartialSubmissionController::SESSION_KEY)) {
+            $request->getSession()->set(PartialSubmissionController::SESSION_KEY, $partial->ID);
+        }
 
-            // Set data record and load the form
-            $this->dataRecord = DataObject::get_by_id($partial->UserDefinedFormClass, $partial->UserDefinedFormID);
-            $this->setFailover($this->dataRecord);
+        // TODO: Recognize visitor with the password
+        // Set data record and load the form
+        $this->dataRecord = DataObject::get_by_id($partial->UserDefinedFormClass, $partial->UserDefinedFormID);
```

Comment:
```suggestion
$record = DataObject::get_by_id($partial->UserDefinedFormClass, $partial->UserDefinedFormID);
$controller = parent::create($record);
$controller->doInit();
```
and replace `$this` with `$controller` seems to work. We don't have to set the failover manually.

Reply:
As discussed on Slack, this will be changed in the next PR.